pkg/controller/gc_controller.go

```diff
@@ -90,70 +99,102 @@ func (c *gcController) Run(ctx context.Context, workers int) error {
-	c.cleanBackups()
+	c.processBackups()
 }
 
-// cleanBackups deletes expired backups.
-func (c *gcController) cleanBackups() {
-	backups, err := c.backupService.GetAllBackups(c.bucket)
-	if err != nil {
-		glog.Errorf("error getting all backups: %v", err)
+// garbageCollectBackup removes an expired backup by deleting any associated backup files (if
+// deleteBackupFile = true), volume snapshots, restore API objects, and the backup API object
+// itself.
+func (c *gcController) garbageCollectBackup(backup *api.Backup, deleteBackupFile bool) {
+	// if the backup includes snapshots but we don't currently have a PVProvider, we don't
+	// want to orphan the snapshots so skip garbage-collection entirely.
+	if c.snapshotService == nil && len(backup.Status.VolumeBackups) > 0 {
+		glog.Warningf("Cannot garbage-collect backup %s because backup includes snapshots and server is not configured with PersistentVolumeProvider",
+			backup.NamespaceAndName())
 		return
 	}
 
-	now := c.clock.Now()
-	glog.Infof("garbage-collecting backups that have expired as of %v", now)
+	// The GC process is primarily intended to delete expired cloud resources (files in object
+	// storage, snapshots). If we fail to delete any of these, we don't delete the Backup API
+	// object so that we don't orphan the cloud resources.
+	deletionFailure := false
 
-	// GC backup files and associated snapshots/API objects. Note that deletion from object
-	// storage should happen first because otherwise there's a possibility the backup sync
-	// controller would re-create the API object after deletion.
-	for _, backup := range backups {
-		if !backup.Status.Expiration.Time.Before(now) {
-			glog.Infof("Backup %s/%s has not expired yet, skipping", backup.Namespace, backup.Name)
-			continue
-		}
-
-		// if the backup includes snapshots but we don't currently have a PVProvider, we don't
-		// want to orphan the snapshots so skip garbage-collection entirely.
-		if c.snapshotService == nil && len(backup.Status.VolumeBackups) > 0 {
-			glog.Warningf("Cannot garbage-collect backup %s/%s because backup includes snapshots and server is not configured with PersistentVolumeProvider",
-				backup.Namespace, backup.Name)
-			continue
+	// If applicable, delete backup file from object storage *before* deleting the API object
+	// because otherwise the backup sync controller could re-sync the backup from object storage.
+	if deleteBackupFile {
+		glog.Infof("Removing backup %s", backup.NamespaceAndName())
+		if err := c.backupService.DeleteBackup(c.bucket, backup.Name); err != nil {
+			glog.Errorf("error deleting backup %s: %v", backup.NamespaceAndName(), err)
+			deletionFailure = true
 		}
+	}
 
-		glog.Infof("Removing backup %s/%s", backup.Namespace, backup.Name)
-		if err := c.backupService.DeleteBackup(c.bucket, backup.Name); err != nil {
-			glog.Errorf("error deleting backup %s/%s: %v", backup.Namespace, backup.Name, err)
+	for _, volumeBackup := range backup.Status.VolumeBackups {
```

Comment:
Let's try to delete the snapshots first, prior to trying to delete the backup file from object storage

Reply:
WDYT about only deleting ark-backup.json if all snapshot deletions and the tarball deletion succeed? this essentially gives us the tombstone functionality, although maybe not an optimal UX. even with this reorder, i realized that a failed snapshot deletion may still get orphaned unless we do something like this.

Reply:
Would we ever have 2 different ways we could end up with just ark-backup.json? e.g. 1) we did a backup and only managed to upload this one file vs 2) we were GCing and were left with just this one?

Reply:
it's theoretically possible, although we do have code that tries to delete the metadata file on create if the tarball upload fails.